### PR TITLE
Pure Node.js? Node.js Without Frameworks? An Overview

### LINKED ISSUE #1192: Pure Node.js? Node.js without frameworks? An overview.
### Brief Summary: 
Frameworks form an intense ecosystem around any Node.js applications. The high chances are that almost every developer will use one or several Node.js packages to develop a web app.

What about the Node.js purest function?  Building Node.js application with no frameworks?  This guide aims to discuss using the Node.js purest functions.

### Key Takeaways: 

Why you should learn to use pure Node.js
The benefits of pure Node.js.
The benefits and drawbacks of using Node.js frameworks as a developer.
The drawbacks of using pure Node.js as a developer.
Build a hello world server with pure Node.js (no frameworks such as Express).

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by kimkimani**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
fixes #1192

### CODE REVIEW

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+Node.js runtime  [download metrics](https://nodejs.org/metrics/) speak for themselves.
+
+![Node.js downoad metrics](/engineering-education/pure-node-js-no-frameworks-or-packages/node-js-downoad-metrics.jpg)
+[Image Source](https://nodejs.org/metrics/)
+
+### An overview
+
+One of the most influencing [reasons for Node.js popularity](/engineering-education/why-node-js-is-popular/) is the availability of diverse [open-source frameworks and libraries](/engineering-education/most-useful-nodejs-packages/).
+
+A package is somebody else code. It is developed to help the community solve day-to-day problems easily and quickly without writing many code setups, especially for a large app.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:33]:
```suggestion
A package is code written by someone else. It is developed to help the community solve day-to-day problems easily and quickly without re-writing code, especially for a large application.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+Open-source codes, frameworks, external libraries, and packages are available for most programming languages. Therefore, you only need basic skills to write your application logic. However, one thing makes Node.js unique. It exposes you to running the core Node.js barebones. With raw Node.js, you can learn and build apps from scratch with no frameworks.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:35]:
```suggestion
Open-source frameworks, external libraries, and packages are available for most programming languages. Therefore, you only need basic skills to write your application logic. However, one thing makes Node.js unique. It exposes you to running the core Node.js barebones. With vanilla Node.js, you can learn and build apps from scratch with no external frameworks.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+Let's take the example of building a Node.js server. In this case, you would use packages such as [Express.js](https://expressjs.com/) to create a server route and manage your HTTP server. Using such a framework is terrific. It makes your server easy to set up, scalable, and easy to manage.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:37]:
```suggestion
Let's take the example of building a Node.js server. In this case, you would use packages such as [Express.js](https://expressjs.com/) to create server routes and manage your HTTP server. Using such a framework is terrific. It makes your server easy to set up, scalable, and easy to manage.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+Libraries make your work efficient by obstructing redundant functionalities concepts. Nevertheless, there is what goes on behind the scenes of these packages. They are developed using pure Node.js. This way, as you implement your HTTP server, the low-level Node.js logic is hidden from you. In such circumstances, your server logic does not directly interact with the pure Node.js.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:39]:
```suggestion
Libraries make your work easier and more efficient by reducing redundant code. Behind the scenes, these libraries are developed using pure Node.js. This way, as you implement your HTTP server, the low-level Node.js logic is hidden from you. In such circumstances, your server logic does not directly interact with vanilla Node.js.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+They are reliable but shrink your learning exposure. Of course, you can't learn everything, but it is important to understand how your system works. Moreover, with that, you start asking yourself the real capability of using Node.js on its own. This drives you to learn the pure Node.js
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:41]:
```suggestion
They are reliable but shrink your learning exposure. Of course, you can't learn everything, but it is important to understand how your system works. Therefore, it is a good experience for a developer to learn vanilla Node.js
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+When learning Node.js, you will typically use packages. As you level up, the urge to learn more might increase, you will realize that you can build APIs and Serves using raw Node.js. No packages or any third-party framework. This way, you utilize the core Node.js runtime code.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:43]:
```suggestion
While learning Node.js, you will typically use packages. As you level up, the urge to learn more might increase, you will realize that you can build APIs and servers using vanilla Node.js, without the use of packages or third-party frameworks. This way, you utilize core Node.js runtime code.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+### Why should you bother learning raw Node.js?
+
+As developers, it's hard to imagine a day without using NPM. A package manager helps you install these packages and use them in your project as if they were build-in. Nevertheless, the NPM ecosystem drives you away and makes it tough to learn a more profound overview of the pure Node.js functionalities.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:47]:
```suggestion
As developers, it's hard to imagine a day without using NPM (Node Package Manager). A package manager helps you install packages and use them in your project. Nevertheless, the NPM ecosystem drives you away and makes it tough to learn the pure Node.js functionalities.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+With packages, every project becomes scalable, fast to build, and easy to manage. This makes it challenging for professionals to dig deeper. Your application development life cycle is reduced to packages names. You call a package function and a set of parameters, and boom, you are done, very easy indeed.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:49]:
```suggestion
With packages, every project becomes scalable, fast to build, and easy to manage. This makes it challenging for professionals to dig deeper. Your application development life cycle is reduced to package names. You call a package function and a set of parameters, and boom, you are done, very easy indeed.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+Learning raw Node.js puts you in a position to develop and contribute to these packages. Packages are built based on pure Node.js. Thus, learning pure Node.js puts you in a position to build a package to help other developers in their development workflow.
+
+Packages are open-source, and frequently there will be issues with the packages, such as bugs. With the knowledge of pure Node.js, you stand in a position to address an issue to solve a bug or improve a package. Packager's code is hosted in GitHub. You can create an issue or a pull request to post your improvement. With that, you become a package contributor, helping the community utilize your skills of using pure Node.js. As a contributor, you understand what happened behind the scene of Node.js packages.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:53]:
```suggestion
Packages are open-source, and frequently there will be issues with the packages, such as bugs. With the knowledge of pure Node.js, you stand in a position to address an issue to solve a bug or improve a package. Package code is hosted in GitHub. You can create an issue or a pull request to post your improvement. With that, you become a package contributor, helping the community utilize your skills of using pure Node.js. As a contributor, you understand what happened behind the scene of Node.js packages.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+We mentioned packages having bugs, so if you develop an app with raw Node.js, How would you run into a bug contributed by a package. Packages have compatibility issues with certain projects due to package versioning specified in the NPM module dependencies. A no package app will not worry about version compatibility. You are the sole manager of your project's functions lifecycle.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:55]:
```suggestion
We mentioned packages having bugs, so if you develop an app with vanilla Node.js, How would you run into a bug contributed by a package. Packages have compatibility issues with certain projects due to package versioning specified in the NPM module dependencies. A no package app will not worry about version compatibility. You are the sole manager of your project's functions lifecycle.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+The biggest drawback of using packages is that a package has a lot of code in it. But you might end only using a small bit of it. Maybe you only need a couple of functions from that library. You will end up having to download the whole package. This will create many unused codes inside your project as the package binaries will be hosted within your project.
+
+When installing a library in your project, its binaries are saved as dependencies in the `node_module` folder. Of course, you need that folder to execute these packages inside your project. Many packages will end up adding many dependencies in the `node_module` folder. That's a couple of bits added to your project.
+
+### A pure Node.js HTTP server
+
+To put the raw Node.js knowledge, I have demonstrated how to create a simple web server. No frameworks, no Node package manager (NPM). Thus no `package.json` file as you have no dependency management—just a text editor and command line and using the functionalities that are core provided by Node.js itself.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:67]:
```suggestion
To put the raw Node.js knowledge, I have demonstrated how to create a simple web server. No frameworks, no Node package manager (NPM). Thus no `package.json` file as you have no dependency management — just a text editor and command line and using the functionalities that are core provided by Node.js itself.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+Import the HTTP module. With HTTP, you can perform different server requests such as GET, POST, PUT, or DELETE. Create a variable to store the returned HTTP instance and include a host and server port. In this case, we are using the localhost that is IP address `127.0.0.1`, listening to port 3000.
+
+```js
+http = require("http");
+const hostname = "127.0.0.1";
+const port = 3000;
+```
+
+- Create an HTTP serve.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:83]:
```suggestion
- Create an HTTP server.
```

File: articles/pure-node-js-no-frameworks-or-packages/index.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+Use the `createServer()` method to create a server instance. It takes a callback function as its second argument. The callback takes arguments, response, and request.
```
Comment (cyberShaw) [inline articles/pure-node-js-no-frameworks-or-packages/index.md:85]:
```suggestion
Use the `createServer()` method to create a server instance. It takes a callback function as its second argument. The callback takes the arguments: response, and request.
```

**Review by cyberShaw — CHANGES_REQUESTED** [review]
Great article, @kimkimani!

But I noticed that this article needs a lot of attention to grammar and readability. Please use a grammar checker to check grammar and make the changes that I've suggested in the comments.

**Comment by kimkimani** [comment]
@cyberShaw can you check this?

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by cyberShaw** [comment]
> @cyberShaw can you check this?

Yep, I'll review the changes today

**Comment by cyberShaw** [comment]
Thank you for making the changes, @kimkimani!

**Comment by hectorkambow** [comment]
@cyberShaw  Is this ready for a final ? Just wanted to double check. 👍

**Comment by cyberShaw** [comment]
> @cyberShaw Is this ready for a final ? Just wanted to double check. 👍

It's ready for a final review, but it would be great if you could check readability once. That's why I haven't added the final review label 😅

**Comment by hectorkambow** [comment]
@kimkimani  Im noticing that the article is a bit choppy and difficult to read. The 1st sentence alone. 

Please be sure to go over the grammar throughly - to avoid either a tiered payout or PR closure. 

Again, an article like this would really benefit from [a professional revision](https://www.fiverr.com/johnbatsoniii/proofread-and-edit-your-writing-for-publication?context_referrer=user_page&ref_ctx_id=9caf75ed-ce79-4880-b2fc-5b09e92e23d6&pckg_id=1&pos=1) - which we have listed a couple options in our [resources page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) with some rather economical prices.

Unfortunately, we can not provide revision services in our program and have to close articles that may take up too much time to polish up. 

🚀  

Thank you

**Comment by kimkimani** [comment]
@hectorkambow Did some extensive grammar and flow checks. It should be good now.

**Comment by cyberShaw** [comment]
> @hectorkambow Did some extensive grammar and flow checks. It should be good now.

I'll do another check

**Comment by cyberShaw** [comment]
Added some edits for readability, in addition to the changes made by the author. Check this article, it's ready for final review. @hectorkambow

**Pull request merged**